### ISSUE TITLE
Can't save image region

### ISSUE THREAD

**Issue by regexgit** [jim-easterbrook/Photini#249]
Set image region, "see modifications":
```
{'Name': {'fr': 'Zone 1'},
 'RegionBoundary': {'rbShape': 'rectangle(0.4334, 0.4859, 0.0831, 0.0795)',
                    'rbUnit': 'relative'}}
```
On save, error:
```
16:47:14: ERROR: photini.metadata: in method 'new_AsciiValue', argument 1 of type 'std::string const &'
Additional information:
Wrong number or type of arguments for overloaded function 'new_AsciiValue'.
  Possible C/C++ prototypes are:
    Exiv2::AsciiValue::AsciiValue()
    Exiv2::AsciiValue::AsciiValue(std::string const &)
    Exiv2::AsciiValue::AsciiValue(Exiv2::Value const &)
Traceback (most recent call last):
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 771, in save
    OK = self._handler_save(
         ^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 752, in _handler_save
    handler.write(name, value)
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 547, in write
    self.set_value(tag, file_value)
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 213, in set_value
    self.set_exif_value(tag, value)
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/exiv2.py", line 601, in set_exif_value
    value = exiv2.AsciiValue(value)
            ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: in method 'new_AsciiValue', argument 1 of type 'std::string const &'
Additional information:
Wrong number or type of arguments for overloaded function 'new_AsciiValue'.
  Possible C/C++ prototypes are:
    Exiv2::AsciiValue::AsciiValue()
    Exiv2::AsciiValue::AsciiValue(std::string const &)
    Exiv2::AsciiValue::AsciiValue(Exiv2::Value const &)
```

Photini 2024.5.0
  Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
  python-exiv2 0.17.0, exiv2 0.28.3
  PySide6 6.7.2, Qt 6.7.2

**Comment by regexgit** [comment]
I realize there is also a problem when loading image:

Image region added by sidecar xmp (with exiv2 -iX), result of `exiv2 -px xxx.jpg`:
```
Xmp.iptcExt.ImageRegion                      XmpText     0  type="Bag"
Xmp.iptcExt.ImageRegion[1]                   XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionName XmpText     8  Region 1
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon XmpText     0  type="Seq"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint XmpText     0  type="Seq"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[1] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[1]/Iptc4xmpExt:PointX XmpText     3  0.3
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[1]/Iptc4xmpExt:PointY XmpText     3  0.4
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[2] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[2]/Iptc4xmpExt:PointX XmpText     3  0.5
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[2]/Iptc4xmpExt:PointY XmpText     3  0.4
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[3] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[3]/Iptc4xmpExt:PointX XmpText     3  0.5
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[3]/Iptc4xmpExt:PointY XmpText     3  0.6
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[4] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[4]/Iptc4xmpExt:PointX XmpText     3  0.3
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint[4]/Iptc4xmpExt:PointY XmpText     3  0.6
```

Error when loading image in photini:
```
16:53:51: ERROR: photini.metadata: 'Iptc4xmpExt:BoundaryPolygon'
Traceback (most recent call last):
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 513, in read
    value = type_.from_exiv2(file_value, tag)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 828, in from_exiv2
    file_value = [cls.item_type.from_exiv2(x, tag) for x in file_value]
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 828, in <listcomp>
    file_value = [cls.item_type.from_exiv2(x, tag) for x in file_value]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 1833, in from_exiv2
    return cls(file_value)
           ^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in __init__
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in <genexpr>
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
                     ^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 1735, in __init__
    super(RegionBoundary, self).__init__(value)
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in __init__
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in <genexpr>
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
                     ^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 692, in get_type
    return cls.item_type[key]
           ~~~~~~~~~~~~~^^^^^
KeyError: 'Iptc4xmpExt:BoundaryPolygon'
16:53:51: ERROR: photini.metadata: 'Iptc4xmpExt:BoundaryPolygon'
Traceback (most recent call last):
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/metadata.py", line 513, in read
    value = type_.from_exiv2(file_value, tag)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 828, in from_exiv2
    file_value = [cls.item_type.from_exiv2(x, tag) for x in file_value]
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 828, in <listcomp>
    file_value = [cls.item_type.from_exiv2(x, tag) for x in file_value]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 1833, in from_exiv2
    return cls(file_value)
           ^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in __init__
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in <genexpr>
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
                     ^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 1735, in __init__
    super(RegionBoundary, self).__init__(value)
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in __init__
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 675, in <genexpr>
    value = dict((k, self.get_type(k, v)(v)) for (k, v) in value.items())
                     ^^^^^^^^^^^^^^^^^^^
  File "/data/applis/venv_python/photini/lib/python3.11/site-packages/photini/types.py", line 692, in get_type
    return cls.item_type[key]
           ~~~~~~~~~~~~~^^^^^
KeyError: 'Iptc4xmpExt:BoundaryPolygon'

```

**Comment by jim-easterbrook** [comment]
To deal with your second point, I'm not familiar with your `BoundaryPolygon[1]/Iptc4xmpExt:PolygonPoint` syntax. The IPTC example files I used when writing Photini have this sort of thing:
```
Xmp.iptcExt.ImageRegion[3]                   XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:Name  LangAlt     1  lang="x-default" Speaker 1
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:OrganisationInImageName XmpBag      1  Organisation name no 1 in region persltr3 (ref2021.1)
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:PersonInImage XmpBag      1  Person name no 1 in region persltr1 (ref2021.1)
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbShape XmpText     7  polygon
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbUnit XmpText     8  relative
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices XmpText     0  type="Seq"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[1] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[1]/Iptc4xmpExt:rbX XmpText     4  0.05
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[1]/Iptc4xmpExt:rbY XmpText     5  0.713
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[2] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[2]/Iptc4xmpExt:rbX XmpText     5  0.148
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[2]/Iptc4xmpExt:rbY XmpText     5  0.041
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[3] XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[3]/Iptc4xmpExt:rbX XmpText     5  0.375
Xmp.iptcExt.ImageRegion[3]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbVertices[3]/Iptc4xmpExt:rbY XmpText     5  0.863
```

I haven't looked into your first question yet, but it would be useful to know what version of python-exiv2 you are using. `python -m photini --version` provides the information I want with any bug report.

**Comment by regexgit** [comment]
It was already at the end of my fist post:
```
Photini 2024.5.0
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
python-exiv2 0.17.0, exiv2 0.28.3
PySide6 6.7.2, Qt 6.7.2
```

**Comment by jim-easterbrook** [comment]
I just tried creating a rectangular region called "Zone 1" in French, and it all seems to work OK:
```
Xmp.iptcExt.ImageRegion                      XmpText     0  type="Bag"
Xmp.iptcExt.ImageRegion[1]                   XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary XmpText     0  type="Struct"
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbShape XmpText     9  rectangle
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbX XmpText     3  0.4
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbY XmpText     3  0.4
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbW XmpText     3  0.2
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbH XmpText     3  0.2
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:RegionBoundary/Iptc4xmpExt:rbUnit XmpText     8  relative
Xmp.iptcExt.ImageRegion[1]/Iptc4xmpExt:Name  LangAlt     1  lang="fr" Zone 1
```

Obviously something in your installation is different. python-exiv2 is the most likely candidate.

**Comment by jim-easterbrook** [comment]
> It was already at the end of my fist post:
> 
> ```
> Photini 2024.5.0
> Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
> python-exiv2 0.17.0, exiv2 0.28.3
> PySide6 6.7.2, Qt 6.7.2
> ```

Sorry I missed it!

**Comment by jim-easterbrook** [comment]
I'm not able to reproduce the problem with either of these:
```
Photini 2024.5.0.post44+g9bce40d0.d20240801
  Python 3.11.9 (main, Apr 18 2024, 16:44:43) [GCC]
  python-exiv2 0.17.0, exiv2 0.28.3
  PySide6 6.7.2, Qt 6.7.2
  system locale en-GB, locales: en-GB en-Latn-GB
  PyEnchant 3.2.2
  ffmpeg version 4.4.4 Copyright (c) 2000-2023 the FFmpeg developers
  styles: Windows, Fusion
  using style: fusion
```

```
Photini 2024.5.0.post49+g43cf97e8.d20240807
  Python 3.6.15 (default, Sep 23 2021, 15:41:43) [GCC]
  python-exiv2 0.17.0, exiv2 0.28.3
  PySide6 6.4.0, Qt 6.4.0
  system locale en-GB, locales: en-GB en-Latn-GB
  PyEnchant 3.2.2
  ffmpeg version 4.4.4 Copyright (c) 2000-2023 the FFmpeg developers
  styles: Windows, Fusion
  using style: fusion
```

**Comment by jim-easterbrook** [comment]
I think the image region stuff is misleading us. Your problem is with an `Exiv2::AsciiValue`. This type is only used for Exif string values, and image regions are stored in XMP, not Exif. I think you might get the same problem if you make some other change to the metadata instead and then try and save the file. Your file might already contain some data that Photini can't handle properly.

If you do have a file that Photini can't modify then I'd find it useful to have a copy of its metadata. You can use `exiv2` to extract all the metadata: `exiv2 -ea file_name`.

**Comment by regexgit** [comment]
You're quite right, it's not related to image regions, all metadata are concerned.

In fact, this concerns all photos that have been processed with my cataloguing tool (Photosupreme on Windows). It's strange because I've often used exiv2 in command line on these photos without any problem.

I'm going to find out which Exif field is messing things up.

**Comment by jim-easterbrook** [comment]
I had similar problems with photos I'd annotated a long time ago with "Windows Pro Photo Tools" (or something like that). It had saved Exif data with non-ascii characters from one of the many Windows character sets. Photini attempts to transcode all string data to utf-8 when it opens a file. (Although the Exif standard says ascii, it's generally accepted to use utf-8 encoding.)

If you could let me have one of your problem files, or the metadata extracted from it, I might be able to make Photini more robust.

**Comment by regexgit** [comment]
You were right again, there are non-ASCII characters.

I ran photini in an IDE with debugger, breakpoint  in exiv2.py , function `set_exif_value` and found:
**tag = Exif.Canon.OwnerName** and  **value=\udca5**

I don't need this metadata so I've removed it and now everything works fine.

One suggestion: add tag and value in photini error

**Comment by jim-easterbrook** [comment]
Improving the error message is difficult without adding exception handlers to a lot of functions, making it harder to read. I'd like to find out why that tag didn't get transcoded, but without an example file there's not much I can do.

**Comment by jim-easterbrook** [comment]
> I'd like to find out why that tag didn't get transcoded.

That's easy - I deliberately don't transcode maker note data, because it's often not Exif standard compliant and I don't want to break anything. I need to find a way to allow it to get written unmodified though.

**Comment by jim-easterbrook** [comment]
Do you remember how you set your camera's owner name? With Canon's EOS utility or something else? I can't get anything but plain ASCII on my 100D's menus.

**Comment by regexgit** [comment]
Late reply because I've done a complete audit of all my photos to detect and correct anomalies, using python scripts with the exiv2 module. I've found a lot of strange things that I haven't finished analysing yet, but the thing that's blocking photini is always invalid UTF values, for example b‘\xa7’ generates an error `’utf-8‘ codec can't encode character “\udca7” in position 0: surrogates not allowed’.
`
I can't say for sure which software was to blame (I tested a lot) but it doesn't matter because now my collection has been cleaned up and Photini is very happy.

Just for info, attached an exiv2 dump of a corrupted photo (it's an .exv but I was forced to rename it to .txt because github doesn't accept this file type).
[CAN_0911.txt](https://github.com/user-attachments/files/16653147/CAN_0911.txt)

**Comment by jim-easterbrook** [comment]
Thanks for the example file. It confirms that my fix 2094b57 works. I get these debugging messages:
```
10:09:42: INFO: photini.exiv2: CAN_0911.exv: detected character set 'iso8859-1'
10:09:42: INFO: photini.exiv2: CAN_0911.exv: transcoded Exif.Canon.FirmwareVersion from 'iso8859-1'
10:09:42: INFO: photini.exiv2: CAN_0911.exv: transcoded Exif.Canon.OwnerName from 'iso8859-1'
```
